arangod/Pregel/GraphStore/GraphLoader.cpp

```diff
@@ -0,0 +1,314 @@
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Markus Pfeiffer
+////////////////////////////////////////////////////////////////////////////////
+
+#include "GraphLoader.h"
+
+#include <memory>
+#include <cstdint>
+
+#include "Pregel/GraphFormat.h"
+#include "Pregel/Algos/ColorPropagation/ColorPropagationValue.h"
+#include "Pregel/Algos/DMID/DMIDValue.h"
+#include "Pregel/Algos/EffectiveCloseness/ECValue.h"
+#include "Pregel/Algos/HITS/HITSValue.h"
+#include "Pregel/Algos/HITSKleinberg/HITSKleinbergValue.h"
+#include "Pregel/Algos/LabelPropagation/LPValue.h"
+#include "Pregel/Algos/SCC/SCCValue.h"
+#include "Pregel/Algos/SLPA/SLPAValue.h"
+#include "Pregel/Algos/WCC/WCCValue.h"
+
+#include "Pregel/Worker/WorkerConfig.h"
+
+#include "Logger/LogMacros.h"
+#include "Scheduler/SchedulerFeature.h"
+#include "Transaction/Helpers.h"
+#include "Transaction/Methods.h"
+#include "Transaction/Options.h"
+#include "Transaction/StandaloneContext.h"
+#include "VocBase/LogicalCollection.h"
+
+#include "ApplicationFeatures/ApplicationServer.h"
+
+#define LOG_PREGEL(logId, level)          \
+  LOG_TOPIC(logId, level, Logger::PREGEL) \
+      << "[job " << config->executionNumber() << "] "
+
+namespace {
+constexpr auto shardError = std::string_view{
+    "Collections need to have the same number of shards,"
+    " use distributeShardsLike"};
+}
+
+namespace arangodb::pregel {
+template<typename V, typename E>
+uint64_t GraphLoader<V, E>::determineVertexIdRangeStart(uint64_t numVertices) {
+  if (arangodb::ServerState::instance()->isRunningInCluster()) {
+    if (config->vocbase()->server().template hasFeature<ClusterFeature>()) {
+      arangodb::ClusterInfo& ci = this->config->vocbase()
+                                      ->server()
+                                      .template getFeature<ClusterFeature>()
+                                      .clusterInfo();
+      return ci.uniqid(numVertices);
+    }
+    ADB_PROD_ASSERT(false) << "ClusterFeature not present in server";
+    return 0;
+  } else {
+    // Warning: when re-parallelised this might become an issue, because
+    // reasons.
+    return 0;
+  }
+}
+
+template<typename V, typename E>
+auto GraphLoader<V, E>::load() -> std::shared_ptr<Quiver<V, E>> {
+  // hold the current position where the ith vertex shard can
+  // start to write its data. At the end the offset should equal the
+  // sum of the counts of all ith edge shards
+
+  // Contains the shards located on this db server in the right order
+  // assuming edges are sharded after _from, vertices after _key
+  // then every ith vertex shard has the corresponding edges in
+  // the ith edge shard
+  std::map<CollectionID, std::vector<ShardID>> const& vertexCollMap =
+      config->vertexCollectionShards();
+  std::map<CollectionID, std::vector<ShardID>> const& edgeCollMap =
+      config->edgeCollectionShards();
+  size_t numShards = SIZE_MAX;
+
+  for (auto const& pair : vertexCollMap) {
```

Comment:
could use structured binding

Reply:
will do this in a further PR; this code is ust copied from previous code to split up the GraphStore into meaningful bits.